options.ts

```diff
@@ -98,36 +78,42 @@ export class OptionsBase {
-		if(optionType === OptionType.Array && typeof result === "string") {
-			return [result];
-		} 
-		
 		return result;
 	}
 
-	public validateOptions(): void {
+	public validateOptions(commandSpecificDashedOptions?: IDictionary<IDashedOption>): void {
+		if(commandSpecificDashedOptions) {
+			this.options = this.globalOptions;
+			_.extend(this.options, commandSpecificDashedOptions);
+			this.setArgv();
+		}
+
 		let parsed = Object.create(null);
-		_.each(_.keys(this.argv), optionName => parsed[optionName] = this.getOptionValue(optionName));
-		_.each(_.keys(parsed), (originalOptionName:string) => {
+		// DO NOT REMOVE { } as when they are missing and some of the option values is false, the each stops as it thinks we have set "return false".
+		_.each(_.keys(this.argv), optionName => {
+			parsed[optionName] = this.getOptionValue(optionName);
+		});
+
+		_.each(parsed, (value:any, originalOptionName:string) => {
+			// when this.options are passed to yargs, it returns all of them and the ones that are not part of process.argv are set to undefined.
+			if(value === undefined) {
+				return;
+			}
+
 			let optionName = this.getCorrectOptionName(originalOptionName);
 
 			if (!_.contains(this.optionsWhiteList, optionName)) {
 				if (!this.isOptionSupported(optionName)) {
-					this.$errors.failWithoutHelp("The option '%s' is not supported. To see command's options, use '$ %s help %s'. To see all commands use '$ %s help'.", originalOptionName, this.$staticConfig.CLIENT_NAME.toLowerCase(), process.argv[2], this.$staticConfig.CLIENT_NAME.toLowerCase());
+					this.$errors.failWithoutHelp(`The option '${originalOptionName}' is not supported. To see command's options, use '$ ${this.$staticConfig.CLIENT_NAME.toLowerCase()} help ${process.argv[2]}'. To see all commands use '$ ${this.$staticConfig.CLIENT_NAME.toLowerCase()} help'.`);
 				} 
-
 				let optionType = this.getOptionType(optionName);
 				let optionValue = parsed[optionName];
-				let parsedOptionType = typeof (optionValue);
-
 				if (_.isArray(optionValue) && optionType !== OptionType.Array) {
 					this.$errors.fail("You have set the %s option multiple times. Check the correct command syntax below and try again.", originalOptionName);
-				} else if (this.doesOptionRequireValue(optionType, parsedOptionType)) {
-					this.$errors.failWithoutHelp("The option '%s' requires a value.", originalOptionName);
 				} else if (optionType === OptionType.String && helpers.isNullOrWhitespace(optionValue)) {
 					this.$errors.failWithoutHelp("The option '%s' requires non-empty value.", originalOptionName);
-				} else if (optionType === OptionType.Boolean && parsedOptionType !== 'boolean') {
-					this.$errors.failWithoutHelp("The option '%s' does not accept values.", originalOptionName);
+				} else if(optionType === OptionType.Array && optionValue.length === 0) {
+					this.$errors.failWithoutHelp("The option '%s' requires non-empty array.", originalOptionName);
```

Comment:
one argument is enough, the code here is for situations like:
`$ appbuilder appmanager upload android --groups`
as groups is defined as array, yargs sets its value to [] . But we need at least one value